### PR TITLE
raftstore: gc log cache eagerly

### PR DESCRIPTION
**Pull request by BusyJay**
For now, log cache will be cleared once it exceeds max cache limit or is compacted. This may not be sufficient enough if log is not compacted in time.

This pr gc cache before they are compacted. For followers, once logs are applied, they won't be accessed again unless there is a campaign. For leaders, when a log is replicated to all nodes and applied, it will not be accessed again. If some nodes are far behind that it has to accept a snapshot, then leader can gc cache after all others have accepted the log. Tests show that memory usage keep low after this pr even in isolation situation.

This pr also makes it easier to fall back to access rocksdb though, especially when election happens.

### CODE REVIEW

File: src/raftstore/store/peer_storage.rs [hunk 1]
```diff
@@ -212,17 +212,13 @@ impl EntryCache {
...
-        };
-
+        let cache_first_idx = self.first_index().unwrap_or(u64::MAX);
         if cache_first_idx > idx {
             return;
         }
         let cache_last_idx = self.cache.back().unwrap().get_index();
         self.cache
-            .drain(..(cmp::min(cache_last_idx, idx) - cache_first_idx) as usize);
+            .drain(..(cmp::min(cache_last_idx + 1, idx) - cache_first_idx) as usize);
```
Comment (overvenus) [inline src/raftstore/store/peer_storage.rs]: Could you add some comments about `+1`?

File: src/raftstore/store/peer_storage.rs [hunk 2]
```diff
@@ -791,6 +789,24 @@ impl PeerStorage {
         self.cache.compact_to(idx);
     }
 
+    pub fn maybe_gc_cache(&mut self, replicated_idx: u64, apply_idx: u64) {
+        if replicated_idx == apply_idx {
```
Comment (hicqu) [inline src/raftstore/store/peer_storage.rs]: Should be `>=`?

Comment (hicqu) [reply src/raftstore/store/peer_storage.rs]: Seems both are ok. But I think `>=` is easier to understand.

**Comment by hicqu** [comment]
LGTM.

File: src/raftstore/store/peer_storage.rs (hunk 2 above)
Comment (BusyJay) [reply src/raftstore/store/peer_storage.rs]: There is compaction rolling all the time, so this strategy is just a supplement for the abnormal cases. One is if the region is inactive, there may be also some logs not compacted, but they are kept in memory. That's the reason why checking equality here. The other is isolation. In this case, `replicated_idx` is not reliable. For example, all follower's progress of a just elected leader will become 0. Just checking equality can reduce the bad decision.

**Comment by BusyJay** [comment]
PTAL

**Comment by BusyJay** [comment]
@overvenus PTAL

**Pull request merged**